pagespeed/system/controller_manager.cc

```diff
@@ -0,0 +1,225 @@
+        errno);
+  } else if (status == 0 /* EOF */) {
+    handler_->Message(kInfo, "Root process exited; controller shutting down.");
+  } else if (status == 1 /* read a byte */) {
+    handler_->Message(
+        kInfo, "Root process is starting a new controller; shutting down.");
+  } else {
+    LOG(FATAL) << "Status of " << status << " doesn't make sense";
+  }
+
+  // TODO(jefftk): Once the controller is doing real work and there's real
+  // cleanup work to do, we can initiate that from here.
+
+  exit(EXIT_SUCCESS);
+
+  // Because the controller is exiting via exit(), the babysitter process will
+  // see that and quit itself instead of restarting it.
+}
+
+void ControllerManager::RunController(SystemRewriteDriverFactory* factory,
+                                      MessageHandler* handler) {
+  handler->Message(kInfo, "Controller running with PID %d", getpid());
+
+  // Would set up gRPC server here and pass control to its event loop.  Instead
+  // just hang out sleeping until the ProcessDeathWatcherThread decides it's
+  // time to exit.
+  while (true) {
+    sleep(60);  // 1m
+  }
+}
+
+void ControllerManager::Daemonize(MessageHandler* handler) {
+  // Make a new session (process group).
+  if (setsid() < 0) {
+    handler->Message(kWarning, "Daemonize: Failed to setsid().");
+  }
+
+  // We need to fork again to make sure there is no session group leader.
+  pid_t pid = fork();
+  CHECK(pid != -1) << "Couldn't fork to daemonize.";
+  if (pid != 0) {
+    exit(EXIT_SUCCESS);
+  }
+
+  // If we keep the current directory we might keep them from being able to
+  // unmount their filesystem.
+  if (chdir("/") < 0) {
+    handler->Message(kWarning, "Daemonize: Failed to chdir(/).");
+  }
+
+  // If we disconnect file descriptors then logging will break, so don't.
+}
+
+void ControllerManager::ForkControllerProcess(
+    SystemRewriteDriverFactory* factory,
+    ThreadSystem* thread_system,
+    MessageHandler* handler) {
+
+  handler->Message(kInfo, "Forking controller process off of %d", getpid());
+
+  // Whenever we fork off a controller we save the fd for a pipe to it.  Then if
+  // we fork off another controller we can write a byte to the pipe to tell the
+  // old controller to clean up and exit.
+  static int controller_write_fd = -1;
+  if (controller_write_fd != -1) {
+    // We already forked off a controller earlier.  Tell it to quit by writing a
+    // byte.  If there's no one still with the pipe open we'll get SIGPIPE and
+    // die horribly, but as long as the babysitter hasn't died that won't
+    // happen.
+    handler->Message(
+        kInfo, "Writing a byte to a pipe to tell the old controller to exit.");
+    ssize_t status;
+    do {
+      status = write(controller_write_fd, "Q", 1);
+    } while (status == -1 && (errno == EAGAIN ||
+                              errno == EINTR));
+    if (status == -1) {
+      handler->Message(kWarning, "killing old controller failed: %s",
```

Comment:
This is a bit racey since it assumes the controller will die immediately. Are you expecting the controller to just die if it can't bind to the port and retry? That's not super nice.